### PR TITLE
Enabling Command Timeout to be specified for Relational Datasources

### PR DESCRIPTION
**Pull request by lukewaters**
Fix for #605

### CODE REVIEW

File: src/EntityFramework.Relational/RelationalConnection.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@ public abstract class RelationalConnection : DataStoreConnection, IDisposable
         private readonly LazyRef<DbConnection> _connection;
         private readonly bool _connectionOwned;
         private int _openedCount;
+        private int? _connectionTimeout;
```
Comment (anpete) [inline src/EntityFramework.Relational/RelationalConnection.cs]: "command"?

Comment (lukewaters) [inline src/EntityFramework.Relational/RelationalConnection.cs]: :+1:

File: src/EntityFramework.Relational/RelationalConnection.cs [hunk 2]
```diff
@@ -74,6 +75,21 @@ public virtual DbConnection DbConnection
...
 
+        public virtual int? CommandTimeout
+        {
+            get { return _connectionTimeout; }
+            set
+            {
+                if (value.HasValue
+                    && value < 0)
+                {
+                    throw new ArgumentException(Strings.InvalidCommandTimeout);
```
Comment (anpete) [inline src/EntityFramework.Relational/RelationalConnection.cs]: Not sure we need to do the range validation here - we are just flowing the value to ADO.

File: src/EntityFramework.Relational/Query/CommandBuilder.cs [hunk 3]
```diff
@@ -36,6 +36,11 @@ public virtual DbCommand Build([NotNull] RelationalConnection connection)
                 command.Transaction = connection.Transaction.DbTransaction;
             }
 
+            if (connection.CommandTimeout != null)
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/CommandBuilder.cs]: What about update pipeline?

File: test/EntityFramework.SqlServer.Tests/SqlServerCommandConfigurationTests.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+using Xunit;
+
+namespace Microsoft.Data.Entity.SqlServer.FunctionalTests
+{
+    public class SqlServerCommandConfigurationTests
+    {
+        public class CommandTimeout
+        {
+            [Fact]
+            public void Default_value_for_CommandTimeout_is_null_and_can_be_changed_including_setting_to_null()
```
Comment (anpete) [inline test/EntityFramework.SqlServer.Tests/SqlServerCommandConfigurationTests.cs]: We need an FT for this - something that ensures that the timeout is propagated to ADO.

File: src/EntityFramework.Relational/Query/CommandBuilder.cs [hunk 5]
```diff
@@ -36,6 +36,11 @@ public virtual DbCommand Build([NotNull] RelationalConnection connection)
...
+            {
+                command.CommandTimeout = (int)connection.CommandTimeout;
```
Comment (anpete) [inline src/EntityFramework.Relational/Query/CommandBuilder.cs]: What happens when CommandTimeout is null here? Is that behavior what we want?

Comment (lukewaters) [inline src/EntityFramework.Relational/Query/CommandBuilder.cs]: Historically, if the user set the CommandTimeout to null, we would use the default value for the DbCommand

File: src/EntityFramework.Relational/Query/CommandBuilder.cs (hunk 3 above)
Comment (lukewaters) [inline src/EntityFramework.Relational/Query/CommandBuilder.cs]: Good catch, I will add it there too

File: src/EntityFramework.Relational/RelationalConnection.cs (hunk 2 above)
Comment (lukewaters) [inline src/EntityFramework.Relational/RelationalConnection.cs]: We did validation here in EF6, I will test without the check when I add the FTs

File: src/EntityFramework.Relational/Query/CommandBuilder.cs (hunk 5 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/CommandBuilder.cs]: Sorry, missed the null check above.

**Comment by lukewaters** [comment]
There are still a few issues with this fix, not ready for the next round of CRing yet

**Comment by lukewaters** [comment]
My changes to this PR are ready for reviewing! @anpete

**Comment by lukewaters** [comment]
Ping...

**Comment by rowanmiller** [comment]
As discussed, please also promote this to a concept on options so that it can come from config (and OnConfiguring). Connection can just pick it up from options when it is created (and it can then be mutated on connection after that). No need to block this PR though, it can be added afterwards.

**Comment by lukewaters** [comment]
My last commit addressed adding CommandTimeout to RelationalDbContextOptions so it can be configured in OnConfiguring.  I would really like to get these changes reviewed and committed soon

File: src/EntityFramework.SqlServer/SqlServerConnection.cs [hunk 6]
```diff
@@ -34,11 +34,15 @@ protected override DbConnection CreateDbConnection()
...
+        public virtual SqlServerConnection CreateMasterConnection()
         {
             var builder = new SqlConnectionStringBuilder { ConnectionString = ConnectionString };
             builder.InitialCatalog = "master";
-            return new SqlConnection(builder.ConnectionString);
+
+            var options = new DbContextOptions();
+            options.UseSqlServer(builder.ConnectionString);
+
+            return new SqlServerConnection(new DbContextService<IDbContextOptions>(() => options), LoggerFactory);
```
Comment (ajcvickers) [inline src/EntityFramework.SqlServer/SqlServerConnection.cs]: Shouldn't the master connection make use of the command timeout?

File: src/EntityFramework.Relational/RelationalDatabase.cs [hunk 7]
```diff
@@ -81,6 +82,12 @@ public virtual Task<bool> HasTablesAsync(CancellationToken cancellationToken = d
             return RelationalDataStoreCreator.HasTablesAsync(cancellationToken);
         }
 
+        public virtual int? CommandTimeout
+        {
+            get { return Connection.CommandTimeout; }
+            set { Connection.CommandTimeout = value; }
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/RelationalDatabase.cs]: The exception message for valid timeout is included above, but there don't seem to be any checks that use it. Can you add one here?

File: src/EntityFramework.Relational/RelationalOptionsExtension.cs [hunk 8]
```diff
@@ -43,6 +44,18 @@ public virtual DbConnection Connection
             }
         }
 
+        public virtual int? CommandTimeout
+        {
+            get { return _commandTimeout; }
+            [param: NotNull]
+            set
+            {
+                Check.NotNull(value, "value");
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/RelationalOptionsExtension.cs]: Add check here as well.

File: src/EntityFramework.Relational/RelationalConnection.cs [hunk 9]
```diff
@@ -74,6 +80,15 @@ public virtual DbConnection DbConnection
...
         public virtual RelationalTransaction Transaction { get; protected set; }
 
+        public virtual int? CommandTimeout
+        {
+            get { return _commandTimeout; }
+            set
+            {
+                _commandTimeout = value;
+            }
+        }
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/RelationalConnection.cs]: It's interesting that you can do:
```
context.Database.AsRelational().CommandTimeout = 10;
```
or
```
context.Database.AsRelational().Connection.CommandTimeout = 10;
```
Do we really need both? We didn't have an abstracted connection class like this in EF6 (other than EntityConnection which was an entirely different beast), but now that we do I wonder if it means that you should always go down to the connection to specify the timeout? This does, however, make it less discoverable. We should discuss briefly with @rowanmiller and/or @divega but I'm okay for this change to go in anyway.

**Comment by ajcvickers** [comment]
:shipit: after adding the argument checks and making the master connection inherit the timeout of the original connection.

File: src/EntityFramework.SqlServer/SqlServerConnection.cs (hunk 6 above)
Comment (lukewaters) [inline src/EntityFramework.SqlServer/SqlServerConnection.cs]: should we also include MaxBatchSize and all of the other configuration that is done on the SqlServerConnection that we are cloning? This makes me feel like we should have an easy way to clone a connection.

Comment (ajcvickers) [inline src/EntityFramework.SqlServer/SqlServerConnection.cs]: Yes. Agreed.

Comment (ajcvickers) [inline src/EntityFramework.SqlServer/SqlServerConnection.cs]: Also see Issue #1406

File: src/EntityFramework.Relational/RelationalDatabase.cs (hunk 7 above)
Comment (lukewaters) [inline src/EntityFramework.Relational/RelationalDatabase.cs]: ok

File: src/EntityFramework.Relational/RelationalOptionsExtension.cs (hunk 8 above)
Comment (lukewaters) [inline src/EntityFramework.Relational/RelationalOptionsExtension.cs]: ok

File: src/EntityFramework.SqlServer/SqlServerConnection.cs (hunk 6 above)
Comment (lukewaters) [inline src/EntityFramework.SqlServer/SqlServerConnection.cs]: I won't address this work as part of this change, but I will make a note on #1406

**Pull request merged**